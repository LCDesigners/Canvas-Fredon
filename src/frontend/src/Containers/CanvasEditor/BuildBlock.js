import React, { Component } from 'react';
import {Segment, Image} from 'semantic-ui-react';
//import uuidv1 from 'uuid/v1';

import PostIt from "../../Components/CanvasEditor/PostIt";
import LinkIt from "../../Components/CanvasEditor/LinkIt";
import {createTextEntry, createReferencePair} from "../../common/model/entry";
import Dashboard from '../../libs/react-dazzle-custom/lib/components/Dashboard';
import {Helper} from "../../Components/Helper";
import stamp from '../../img/postage-stamp.png';
import { SharedSnackbarProvider } from './SharedSnackbarProvider';
import { changeEntryTags } from './actions';

import './style.css';

/**
 * @class BuildBlock : react component - a semantic BuildingBlock shown in a canvas
 * a BuildBlock can have different types : plain/target/link
 *  -> specifies the type and behavior of the child entries
 */
class BuildBlock extends Component{

  /**
   * specifies the default propTypes
   */
  static propTypes = {

  };

  /**
   * constructor
   * @param {*} props all handed props
   */
  constructor(props) {
    super(props);
    this.buildStructure = this.buildStructure.bind(this);
    this.entryReps={};
    this.state = {...this.state||{}}
  }

  componentDidMount(){
    this.setState({...this.state||{} });
    window.matchMedia("print").addListener(this.printListener);
  }

  componentWillUnmount(){
    window.matchMedia("print").removeListener(this.printListener);
  }

  /**
   * generates the displayed layout structure for react-dazzle based on its context
   */
    buildStructure(){
      /*
        first, check if block is locked by actions from other users
        then, apply the editability of the whole canvas with logical and (maybe you are in
        history mode or have no rights to do something)
      */
      let editable = true;
      if(this.props.lockings !== undefined && this.props.lockings.indexOf(this.props.block._id) > -1){
        editable = false;
      }
      editable = editable && this.props.editable;

      /*
        by react-dazzle required structure
        see npm react-dazzle for more information
      */

      let structure = {
        widgets: {
          PostIt: {
            type: PostIt,
            title: 'Entry'
          },
          LinkIt: {
            type: LinkIt,
            title: 'Entry'
          }
        },
        layout: {
          rows: [{
            columns: [{
              className: 'col-md-12 col-sm-12 col-xs-12',
              widgets: [],
            }]
          }]
        }
      };

      /*
            create different entries depending on the type of the BuildBlock
            */

      // TODO this is a hack because I fail to properly remove the TargIt
      if (this.props.block.buildingBlockType === "target") {
        this.props.block.buildingBlockType = "data";
      }
      var postItClass = "PostIt";
      if (this.props.block.buildingBlockType === "link") {
        postItClass = "LinkIt";
      }

      Object.values(this.props.block.entries).forEach((entry,index) => {
        structure.layout.rows[0].columns[0].widgets.push({key: postItClass, title: entry.content.title ,text: entry.content.text,
          entyID: entry._id, blockID: this.props.block._id, canvasID:this.props.canvasid,
          project_id: this.props.project_id, model: entry, callback: this.onChangeEntry.bind(this), onKeyDown: this.onKeyDown.bind(this), editable: editable,
          openSettings : this.props.openSettings,
          globalTags:this.props.globalTags,
          enableMarkdown:this.props.enableMarkdown,
          ref: (entryRep)=>{
            if(entryRep){
              this.entryReps[entry._id]=entryRep;
            }else{
              delete this.entryReps[entry._id];
            }}});
      });
      return structure;
    }

    /**
     * onRemoveEntry
     * @param {*} e - entry wich should be removed
     * if entry is of type plain, it simply gets removed
     * if type link, the entry and its linked target entry are removed both
     * entrys of type link cannot be removd on their own
     *  -> pushes callback to parent
     */
    onRemoveEntry(e){
      let entry = Object.values(this.props.block.entries)[e.index]
      entry.index = e.index;
      switch(this.props.block.buildingBlockType){
      case 'data' :
        this.props.actions.onRemoveEntry(entry, this.props.block._id);
        break;
      case 'link' :
        //let target = {};
        //target.index = en.index;
        //this.props.actions.onRemoveEntry(target, this.props.refblock._id);
        this.props.actions.onRemoveEntry(entry, this.props.block._id);
      break;
      default:
        return;
      }
    }

    createEmptyEntry(){
      let entry = createTextEntry();
      entry.content.text = "";
      switch(this.props.block.buildingBlockType){
        case 'data':
          entry.type="plain";
          break;
        case 'link':
          entry.type="link";
        break;
        default: return;
      }
      return entry;
    }

    addTextEntry(text="",title="default",bindFocus=false){
      let entry = createTextEntry(text);
      this.props.actions.onAddEntry(entry,this.props.block._id);
      if(bindFocus) window.setTimeout(this.focusEntry.bind(this),10,entry._id);
    }

    addLinkEntry(text="",title="default",bindFocus=false){
      // eslint-disable-next-line
      let {linkIt, canvas} = createReferencePair(text, title, this.props.canvasid)
      let blocks={
        [this.props.block._id]:{
          [linkIt._id]:linkIt
        },
        /*
        [this.props.refblock._id]:{
          [targIt._id]:targIt
        }
        */
      }

      this.props.actions.addEntryCanvas(this.props.canvasid,blocks,{[canvas._id]:canvas});
      if(bindFocus) window.setTimeout(this.focusEntry.bind(this),10,linkIt._id);
    }

    addEntry(text="",title="default", bindFocus=false){
      if ((typeof text)!=="string") throw new Error("Text has to be a string! (got "+JSON.stringify(text)+")");
      if(this.props.block.buildingBlockType === "data"){
        this.addTextEntry(text,title,bindFocus);
      }else if(this.props.block.buildingBlockType === "link"){
        this.addLinkEntry(text,title,bindFocus);
      }
    }

    onKeyDown(event, model){
      event.persist();
      let {shiftKey,key}=event;
      if(!shiftKey){
        let {selectionStart,selectionEnd}=event.target;
        let isSelection=(selectionEnd!==selectionStart);
        switch(key){
          case "Enter":
            if(isSelection) return;
            if(selectionStart!==model.content.text.length) return;
            this.addEntry("",null,true);
            break;
          default:
            return;
        }
        event.preventDefault();
      }
  }

  reformatImportText(input){
    //Better be safe than sorry and not autoformat anything with a "?", might be something like "spam.stuff.io/autosubscribe?deref=friendly.png"
    if(input.startsWith("data:image")||/^http(?:s?):\/\/[\d\w]+\.[^?]+\/[^?]+\.(jpg|gif|png|bmp)$/.test(input))
      return `![](${input})`
    return input;
  }


  onDragDrop(event, entryID=undefined){
    event.stopPropagation();
    event.preventDefault();
    let textContent = event.dataTransfer.getData("text");
    
    let jsonContent = null;
    let source;
    let textMarkers;
    let targetPosition=-1;
    if (entryID) {
      targetPosition=Object.keys(this.props.block.entries).indexOf(entryID);
      const newTags =[this.props.draggedTagId];
      const delTags = [];
      this.props.dispatch(changeEntryTags(entryID, this.props.block._id, this.props.canvasid, newTags, delTags));
    }
    try {
      jsonContent = JSON.parse(event.dataTransfer.getData("json"));
      source = jsonContent.source;
      textMarkers=jsonContent.textMarkers;
    } catch (error) {
      source=null;
      textMarkers=null;
    }
    if (source) {
      if(textMarkers){
        let {start,end}=textMarkers;
        let sourceEntry=this.props.findEntryByIDs(source.entry,source.block,source.canvas);
        
        if((!sourceEntry)||(sourceEntry.content.text.substring(start,end)!==textContent)){
          this.addEntry(textContent);
          console.log("Could not verify text-source, creating new entry instead");
        }else if(this.props.block.buildingBlockType!=="data"){
          this.addEntry(textContent);
          console.log("Cannot extract to link block, creating new entry instead");
        }else{
          let newText=sourceEntry.content.text
          newText=newText.substring(0,start)+newText.substring(end);
          let newEntry=createTextEntry(textContent);
          this.props.onChangeAdd([{
              ...sourceEntry,content:{...sourceEntry.content,text:newText}},
              source.block,source.canvas],
            [newEntry,this.props.block._id,this.props.canvasid]);
        }
      }else{
        this.pullEntry(source,targetPosition);
      }
    } else if (textContent){
      this.addEntry(this.reformatImportText(textContent));
    }
  }

  onDragOver(event, entryID=undefined){
    event.preventDefault();
    //event.persist();
    event.dataTransfer.dropEffect = "move"
  }

    focusEntry(entryID){
      if(this.entryReps[entryID]){
        if(this.entryReps[entryID].onFocus)
          this.entryReps[entryID].onFocus();
        if(this.entryReps[entryID].textArea){
          this.entryReps[entryID].textArea.focus();
          return true;
        }
      }
      return false;
    }

    pullEntry(source,targetPosition=-1){
      //console.log('pullEntry is called');
      let {canvas,block,entry}=source;
      if(this.props.canvasid !== canvas) throw new Error("Inter-canvas drag&drop is not supported yet");//TODO
      if(targetPosition!==-1)
        this.props.moveEntry(block,entry,this.props.block._id,targetPosition);
      else
        this.props.moveEntry(block,entry,this.props.block._id);
    }

    /**
     * onChangeEntry
     * @param {*} e - entry wich content changed
     *  -> pushes callback to parent
     */
    onChangeEntry(e){
      this.props.actions.onChangeEntry(e, this.props.block._id);
    }

    /**
     * onReferenceCall
     * @param {*} linkIt - link entry wich is clicked
     * @deprecated
     *  -> pushes callback to parent
     */
    onReferenceCall(linkIt){
      this.props.actions.onReferenceCall(linkIt);
    }

    /**
     * render : from react.component : renders the BuildBlock
     */
    render() {

      //first, check if block is editable
      let editable = true;
      let canAdd = true;
      //maybe you have no rights or are in history modes
      if(this.props.editable === false){
        editable = false;
      }
      //mybe the block is locked by actions from other users
      if(this.props.lockings !== undefined && this.props.lockings.indexOf(this.props.block._id) > -1){
        editable = false;
      }

      //build structure (layout) fro react-dazzle lib
      let structure = this.buildStructure();
      let blockTitle = this.props.block.title;

      return(
        <Segment className={this.props.className} 
                  onDragOver={this.onDragOver.bind(this)} 
                  onDrop = {this.onDragDrop.bind(this)} >
        <div style={this.props.style}>
        <span>

        <h3>{this.props.block.title} 
          <SharedSnackbarProvider>
            {this.props.help && (<Helper topic={this.props.help}/>)}
          </SharedSnackbarProvider>
        </h3>
        </span>
        {blockTitle === "Letter to Grandma" ? (
          <div>
            <Image src={stamp}  style={{width: 66, height: 64, float:'right'}}  size="small" />
          </div>
        ) : ( <div/>
        )}
        <Dashboard
          type = {this.props.type}
          widgets = {structure.widgets}
          layout = {structure.layout}
          editable = {editable}
          onRemove = {this.onRemoveEntry.bind(this)}
          onAdd = {this.addEntry.bind(this)}
          onChange ={this.onChangeEntry.bind(this)}
          entryDragOver = {this.onDragOver.bind(this)}
          entryDrop = {this.onDragDrop.bind(this)}
          //dragable = {false}
          canAdd = {canAdd}
        /></div>
        </Segment>
      );
    }

}

export default BuildBlock;
