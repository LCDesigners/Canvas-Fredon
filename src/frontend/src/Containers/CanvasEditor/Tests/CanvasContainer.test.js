import React from 'react';
import renderer from 'react-test-renderer';
import 'babel-polyfill';

import BuildBlock from '../BuildBlock';

// mock for matchMedia from https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function
window.matchMedia = window.matchMedia || function() {
  return {
    matches : false,
    addListener : function() {},
    removeListener: function() {}
  };
};

const entry1 = {_id: 'e1', buildingBlockType: 'data', content: {text: 'eergresf'}};
const entry2 = {_id: 'e2', buildingBlockType: 'data', content: {text: 'frgersdg'}};
const entry3 = {_id: 'e3', buildingBlockType: 'data', content: {text: undefined }};
const entry4 = {_id: 'e4', buildingBlockType: 'link', content: {text: undefined, target: 'e5', reference: 'somecanvasid'}};
const entry5 = {_id: 'e5', buildingBlockType: 'target', content: {text: 'fgergs'}};

const b0 = {_id: 'b0', title: 'Key Partners', buildingBlockType: 'data', layoutEntry:{}, entries:{e1: entry1}};
const b1 = {_id: 'b1', title: 'Key Activities', buildingBlockType: 'data', layoutEntry:{}, entries:{e1: entry1, e2: entry2}};
const b2 = {_id: 'b2', title: 'Key Resources', buildingBlockType: 'data', layoutEntry:{}, entries:{}};
const b3 = {_id: 'b3', title: 'Customer Relationships', buildingBlockType: 'data', layoutEntry:{}, entries:{e1: entry1, e2: entry2, e3: entry3}};
const b4 = {_id: 'b4', title: 'Channels', buildingBlockType: 'data', layoutEntry:{}, entries:{}};
const b5 = {_id: 'b5', title: 'Customer Segments', buildingBlockType: 'target', layoutEntry:{}, entries:{e5: entry5}};

const blockDef1 = {
    block: b0,
    buildingBlockType: "data",
    lockings: [],
    editable: true,
    project_id: 'pid1',
    actions: {onChangeEntry : () => {}}
}

const blockDef2 = {
    block: b3,
    buildingBlockType: "data",
    lockings: [],
    editable: true,
    project_id: 'pid1',
    actions: {onChangeEntry : () => {}}
}

const blockDef3 = {
    block: b5,
    buildingBlockType: "target",
    lockings: [],
    editable: true,
    project_id: 'pid1',
    actions: {onChangeEntry : () => {}}
}

test('BuildBlock should display the expected state', () => {

    let component = renderer.create(
      <BuildBlock
        block={blockDef1.block}
        lockings={blockDef1.lockings}
        editable={blockDef1.editable}
        project_id={blockDef1.project_id}
        actions={blockDef1.actions}>
        >
      </BuildBlock>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    component = renderer.create(
        <BuildBlock
          block={blockDef2.block}
          lockings={blockDef2.lockings}
          editable={blockDef2.editable}
          project_id={blockDef2.project_id}
          actions={blockDef2.actions}>
          >
        </BuildBlock>
      );

      tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      component = renderer.create(
        <BuildBlock
          block={blockDef3.block}
          lockings={blockDef3.lockings}
          editable={blockDef3.editable}
          project_id={blockDef3.project_id}
          actions={blockDef3.actions}>
          >
        </BuildBlock>
      );

      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
});
