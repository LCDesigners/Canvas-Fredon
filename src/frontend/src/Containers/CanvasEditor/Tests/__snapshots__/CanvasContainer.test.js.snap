// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildBlock should display the expected state 1`] = `
<div
  className="ui segment"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div>
    <span>
      <h3>
        Key Partners
           
      </h3>
    </span>
    <div>
      <div>
        <div
          className="row"
        >
          <div
            className="col-md-12 col-sm-12 col-xs-12 editable-column"
          >
            <div
              className="add-widget-button"
              onClick={[Function]}
            >
              <span
                className="add-widget-link"
              >
                +
              </span>
            </div>
            <div>
              <div
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <div
                  className="defaultWidgetFrame"
                >
                  <div
                    className="defaultWidgetFrameHeader"
                  >
                    <span
                      className="title"
                    >
                      Entry
                    </span>
                    <span
                      className="remove"
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="remove icon removeIcon"
                      />
                    </span>
                  </div>
                  <span
                    draggable="true"
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <span
                      className="tagbar"
                    >
                      <svg
                        cursor="pointer"
                        height={20}
                        onClick={[Function]}
                        preserveAspectRatio="xMinYMin"
                        width="8"
                      />
                    </span>
                    <textarea
                      className="textarea"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      rows="1"
                      style={
                        Object {
                          "resize": "none",
                        }
                      }
                      value="eergresf"
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BuildBlock should display the expected state 2`] = `
<div
  className="ui segment"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div>
    <span>
      <h3>
        Customer Relationships
           
      </h3>
    </span>
    <div>
      <div>
        <div
          className="row"
        >
          <div
            className="col-md-12 col-sm-12 col-xs-12 editable-column"
          >
            <div
              className="add-widget-button"
              onClick={[Function]}
            >
              <span
                className="add-widget-link"
              >
                +
              </span>
            </div>
            <div>
              <div
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <div
                  className="defaultWidgetFrame"
                >
                  <div
                    className="defaultWidgetFrameHeader"
                  >
                    <span
                      className="title"
                    >
                      Entry
                    </span>
                    <span
                      className="remove"
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="remove icon removeIcon"
                      />
                    </span>
                  </div>
                  <span
                    draggable="true"
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <span
                      className="tagbar"
                    >
                      <svg
                        cursor="pointer"
                        height={20}
                        onClick={[Function]}
                        preserveAspectRatio="xMinYMin"
                        width="8"
                      />
                    </span>
                    <textarea
                      className="textarea"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      rows="1"
                      style={
                        Object {
                          "resize": "none",
                        }
                      }
                      value="eergresf"
                    />
                  </span>
                </div>
              </div>
              <div
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <div
                  className="defaultWidgetFrame"
                >
                  <div
                    className="defaultWidgetFrameHeader"
                  >
                    <span
                      className="title"
                    >
                      Entry
                    </span>
                    <span
                      className="remove"
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="remove icon removeIcon"
                      />
                    </span>
                  </div>
                  <span
                    draggable="true"
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <span
                      className="tagbar"
                    >
                      <svg
                        cursor="pointer"
                        height={20}
                        onClick={[Function]}
                        preserveAspectRatio="xMinYMin"
                        width="8"
                      />
                    </span>
                    <textarea
                      className="textarea"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      rows="1"
                      style={
                        Object {
                          "resize": "none",
                        }
                      }
                      value="frgersdg"
                    />
                  </span>
                </div>
              </div>
              <div
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <div
                  className="defaultWidgetFrame"
                >
                  <div
                    className="defaultWidgetFrameHeader"
                  >
                    <span
                      className="title"
                    >
                      Entry
                    </span>
                    <span
                      className="remove"
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="remove icon removeIcon"
                      />
                    </span>
                  </div>
                  <span
                    draggable="true"
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                  >
                    <span
                      className="tagbar"
                    >
                      <svg
                        cursor="pointer"
                        height={20}
                        onClick={[Function]}
                        preserveAspectRatio="xMinYMin"
                        width="8"
                      />
                    </span>
                    <textarea
                      className="textarea"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onInput={[Function]}
                      onKeyDown={[Function]}
                      rows="1"
                      style={
                        Object {
                          "resize": "none",
                        }
                      }
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BuildBlock should display the expected state 3`] = `
<div
  className="ui segment"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div>
    <span>
      <h3>
        Customer Segments
           
      </h3>
    </span>
    <div>
      <div>
        <div
          className="row"
        >
          <div
            className="col-md-12 col-sm-12 col-xs-12"
          >
            <div>
              <div
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <div
                  className="defaultWidgetFrame"
                >
                  <div
                    className="defaultWidgetFrameHeader"
                  >
                    <span
                      className="title"
                    >
                      Entry
                    </span>
                  </div>
                  <div
                    className="x-postit"
                    style={
                      Object {
                        "borderColor": "#a9d62c",
                      }
                    }
                  >
                    <span
                      draggable="true"
                      onDragOver={[Function]}
                      onDragStart={[Function]}
                      onDrop={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                    >
                      <span
                        className="tagbar"
                      >
                        <svg
                          cursor="pointer"
                          height={20}
                          onClick={[Function]}
                          preserveAspectRatio="xMinYMin"
                          width="8"
                        />
                      </span>
                      <textarea
                        className="textarea"
                        disabled={false}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onInput={[Function]}
                        onKeyDown={[Function]}
                        rows="1"
                        style={
                          Object {
                            "resize": "none",
                          }
                        }
                        value="fgergs"
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
